Update Using AWS Inferentia to run HuggingFace TorchScript model

kct22aws opened the pull request:
# What does this PR do?
This PR appended a new section "Using HuggingFace TorchScript model in AWS Inf1 using Neuron SDK" to the TorchScript section of the HuggingFace transformer documentation.

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)
Link to issue: https://github.com/huggingface/transformers/issues/14425


## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link 
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?
@philschmid, @LysandreJik

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

kct22aws committed:
Update Using AWS Inferentia to run HuggingFace TorchScript model

kct22aws commented:
This is a PR for updating TorchScript section with guidance on how to use AWS Inf1 instances to run HuggingFace transformer models. @philschmid, @LysandreJik I look forward to your reviews.

philschmid commented on docs/source/serialization.rst:
```diff
@@ -430,3 +430,67 @@ Using the traced model for inference is as simple as using its ``__call__`` dund
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Models with transformer based architecture, i.e., 
+`BERT (Bidirectional Encoder Representations from Transformers) <https://huggingface.co/bert-base-uncased>`_ or its variants
+such as `distilBERT <https://huggingface.co/transformers/v4.0.1/model_doc/distilbert.html>`_ 
+and `roBERTa <https://huggingface.co/roberta-base>`_ will run best in Inf1 for non-generative tasks such as Extractive Question Answering, Sequence Classification, NER. 
```
We should align the ref links to those models. I think adding the `docs` link for all 3 is probably better than the models weights on the hub or what do you think?

kct22aws replied:
Agreed and resolved. .DS_Store file removed, and links are all referring to the master docs links. Rebased and resubmitted to a new PR: https://github.com/huggingface/transformers/pull/14982

philschmid commented on docs/source/serialization.rst:
```diff
@@ -430,3 +430,67 @@ Using the traced model for inference is as simple as using its ``__call__`` dund
+such as `distilBERT <https://huggingface.co/transformers/v4.0.1/model_doc/distilbert.html>`_ 
+and `roBERTa <https://huggingface.co/roberta-base>`_ will run best in Inf1 for non-generative tasks such as Extractive Question Answering, Sequence Classification, NER. 
+Alternatively, text generation tasks can be adapted to run on Inf1, according to this tutorial using `MarianMT <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/src/examples/pytorch/transformers-marianmt.html>`_.
+More information about models that perform well out of the box on Inferentia can be found in the `Model Architecture Fit section of the Neuron documentation <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-guide/models/models-inferentia.html#models-inferentia>`_.
```
```suggestion
More information about models that can be converted out of the box for Inf1 can be found in the `Model Architecture Fit section of the Neuron documentation <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-guide/models/models-inferentia.html#models-inferentia>`_.
```

philschmid commented on docs/source/serialization.rst:
```diff
@@ -430,3 +430,67 @@ Using the traced model for inference is as simple as using its ``__call__`` dund
+Using HuggingFace TorchScript model in AWS Inf1 using Neuron SDK
+-----------------------------------------------------------------------------------------------------------------------
+
+AWS introduced `Inf1 <https://aws.amazon.com/ec2/instance-types/inf1/>`_ for low cost, high performance machine learning model inference tasks in AWS cloud. 
```
We should think about calling inferentia: `inf1` or `inferentia` in the whole document and not mix them up. This can confuse none aws experienced people

kct22aws replied:
Agreed and all should be Inf1. Changes made in new commit.

kct22aws replied:
Appreciate the suggestion and changes are made where appropriate. We used the term 'Inf1' to refer to EC2 instance type, and 'Inferentia' to refer to the chip name. Rebased and resubmit in a new PR: https://github.com/huggingface/transformers/pull/14982

philschmid commented on docs/source/serialization.rst:
```diff
@@ -430,3 +430,67 @@ Using the traced model for inference is as simple as using its ``__call__`` dund
+Models with transformer based architecture, i.e., 
+`BERT (Bidirectional Encoder Representations from Transformers) <https://huggingface.co/bert-base-uncased>`_ or its variants
+such as `distilBERT <https://huggingface.co/transformers/v4.0.1/model_doc/distilbert.html>`_ 
+and `roBERTa <https://huggingface.co/roberta-base>`_ will run best in Inf1 for non-generative tasks such as Extractive Question Answering, Sequence Classification, NER. 
```
```suggestion
and `roBERTa <https://huggingface.co/roberta-base>`_ will run best in Inf1 for non-generative tasks such as Extractive Question Answering, Sequence Classification, Token Classification. 
```

philschmid reviewed (commented):
Thank you for the PR this looks great and will be a good addition! 
I added a few comments. 
Additionally, can you remove the `.DS_STORE` files for your commits?

kct22aws committed:
Update docs/source/serialization.rst

Co-authored-by: Philipp Schmid <32632186+philschmid@users.noreply.github.com>

kct22aws committed:
Update docs/source/serialization.rst

Co-authored-by: Philipp Schmid <32632186+philschmid@users.noreply.github.com>

kct22aws committed:
Remove .DS_Store files

kct22aws committed:
Revise starting section title and correct reference to instance or chip

kct22aws committed:
Change perform well to can be converted

Signed-off-by: K.C. Tung <kctung@amazon.com>

philschmid commented:
@kct22aws could you rebase your branch with the master so only your changes are shown? An in addition the `check_code_quality` CI is failing. 
Could you run `make style` to fix this?

kct22aws committed:
Verify to pass make style

kct22aws commented:
> 
Hi @philschmid I committed again after a rebase, and also passed the style and format checks.

philschmid commented on docs/source/serialization.rst:
```diff
@@ -427,6 +427,91 @@ Using a traced model for inference
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Transformers Models based on the `BERT (Bidirectional Encoder Representations from Transformers)
+<https://huggingface.co/bert-base-uncased>`_ architecture, or its variants such as `distilBERT
```
Can you please align those documentation links to the newest documentation, e.g. `distilbert` -> https://huggingface.co/docs/transformers/master/model_doc/distilbert

philschmid commented on docs/source/serialization.rst:
```diff
@@ -427,6 +427,91 @@ Using a traced model for inference
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Transformers Models based on the `BERT (Bidirectional Encoder Representations from Transformers)
+<https://huggingface.co/bert-base-uncased>`_ architecture, or its variants such as `distilBERT
+<https://huggingface.co/transformers/v4.0.1/model_doc/distilbert.html>`_ and `roBERTa
+<https://huggingface.co/roberta-base>`_ will run best on Inf1 for non-generative tasks such as Extractive Question
```
Can you please align those documentation links to the newest documentation, e.g. `distilbert` -> https://huggingface.co/docs/transformers/master/model_doc/distilbert

kct22aws replied:
Thank you and it is changed according to your suggestion.

philschmid commented on docs/source/serialization.rst:
```diff
@@ -427,6 +427,91 @@ Using a traced model for inference
+
+    traced_model(tokens_tensor, segments_tensors)                   
+
+Deploying HuggingFace TorchScript models on AWS using the Neuron SDK
```
```suggestion
Converting HuggingFace Transformers to AWS Neuron compatible models
```

philschmid commented on docs/source/serialization.rst:
```diff
@@ -427,6 +427,91 @@ Using a traced model for inference
+Dependencies
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Using AWS Neuron to run inference workloads requires the following dependencies and environment:
```
```suggestion
Using AWS Neuron to convert models requires the following dependencies and environment:
```

philschmid commented on docs/source/serialization.rst:
```diff
@@ -427,6 +427,91 @@ Using a traced model for inference
+
+Using AWS Neuron to run inference workloads requires the following dependencies and environment:
+
+1. An `Amanzon EC2 Inf1 <https://aws.amazon.com/ec2/instance-types/inf1/>`_ instance in AWS cloud.
```
```suggestion
```

philschmid commented on docs/source/serialization.rst:
```diff
@@ -427,6 +427,91 @@ Using a traced model for inference
+Using AWS Neuron to run inference workloads requires the following dependencies and environment:
+
+1. An `Amanzon EC2 Inf1 <https://aws.amazon.com/ec2/instance-types/inf1/>`_ instance in AWS cloud.
+2. A `Neuron SDK environment
```
```suggestion
* A `Neuron SDK environment
```

philschmid commented on docs/source/serialization.rst:
```diff
@@ -427,6 +427,91 @@ Using a traced model for inference
+   <https://docs.aws.amazon.com/dlami/latest/devguide/tutorial-inferentia-launching.html>`_
+
+
+Running Neuron traced TorchScript in Inf1
```
```suggestion
Converting a Model for AWS Neuron
```

philschmid reviewed (commented):
I think we need to change the wording from "running" inference to "converting/exporting" since this the topic of the page.

kct22aws committed:
Update docs/source/serialization.rst

Modify per reviewer suggestion

Co-authored-by: Philipp Schmid <32632186+philschmid@users.noreply.github.com>

kct22aws committed:
Update docs/source/serialization.rst

Change per reviewer suggestion: 'run inference workload' to 'convert models'

Co-authored-by: Philipp Schmid <32632186+philschmid@users.noreply.github.com>

kct22aws committed:
Update docs/source/serialization.rst

Accept reviewer suggestion to remove this sentence

Co-authored-by: Philipp Schmid <32632186+philschmid@users.noreply.github.com>

kct22aws committed:
Update docs/source/serialization.rst

Accept reviewer suggestion to use asterick

Co-authored-by: Philipp Schmid <32632186+philschmid@users.noreply.github.com>

kct22aws committed:
Update docs/source/serialization.rst

Accept reviewer suggestion to change 'Running Neuron traced TorchScript in Inf1' to 'Converting a Model for AWS Neuron'

Co-authored-by: Philipp Schmid <32632186+philschmid@users.noreply.github.com>

kct22aws committed:
Accept all reviewer suggestions and pass make style

kct22aws commented:
Hi @philschmid all your suggestions are implemented and commit passed checks.

philschmid commented on docs/source/serialization.rst:
```diff
@@ -427,6 +427,90 @@ Using a traced model for inference
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Transformers Models based on the `BERT (Bidirectional Encoder Representations from Transformers)
+<https://huggingface.co/bert-base-uncased>`_ architecture, or its variants such as `distilBERT
```
This link is still wrong

kct22aws replied:
HI @philschmid thank you for the suggestion, the link should be consistent with that of other models now. I updated BERT link to https://huggingface.co/docs/transformers/master/model_doc/bert

kct22aws committed:
Update link for BERT to master doc

kct22aws commented:
Hi @philschmid @LysandreJik all reference links to BERT and BERT models are updated and should be consistent now.

philschmid commented:
Thank you @kct22aws. Could you rebase and resolve the conflicts one last time then we could merge it.

kct22aws commented:
Hi Philipp,
I rebased and checkout a new branch that is consistent with current main branch. My changes are now in serialization.mdx. A new PR is in https://github.com/huggingface/transformers/pull/14982
Hope this works.
Regards,
KC

From: Philipp Schmid ***@***.***>
Reply-To: huggingface/transformers ***@***.***>
Date: Wednesday, December 29, 2021 at 3:29 AM
To: huggingface/transformers ***@***.***>
Cc: "Tung, KC" ***@***.***>, Mention ***@***.***>
Subject: Re: [huggingface/transformers] Update Using AWS Inferentia to run HuggingFace TorchScript model (PR #14771)


Thank you @kct22aws<https://github.com/kct22aws>. Could you rebase and resolve the conflicts one last time then we could merge it.

—
Reply to this email directly, view it on GitHub<https://github.com/huggingface/transformers/pull/14771#issuecomment-1002480962>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AU3SAP7GP66DJNRYY63SWQLUTLIFNANCNFSM5KCJLPKQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

philschmid commented:
* https://github.com/huggingface/transformers/pull/14982
is merged
